src/inversion/inversion.f90

```diff
@@ -24,49 +24,64 @@ subroutine vor2vel
+!             ! Enforce solenoidality
+!             ! A, B, C are vorticities
+!             ! D = B_x - A_y; E = C_z
+!             ! A = k2l2i * (E_x + D_y) and B = k2l2i * (E_y - D_x) --> A_x + B_y + C_z = zero
+!             call diffx(svor(:, :, :, 2), as) ! as = B_x
+!             call diffy(svor(:, :, :, 1), bs) ! bs = A_y
+!             !$omp parallel workshare
+!             ds = as - bs                     ! ds = D
+!             cs = svor(:, :, :, 3)
+!             !$omp end parallel workshare
+!             call field_combine_semi_spectral(cs)
+!             call central_diffz(cs, es)                     ! es = E
+!             call field_decompose_semi_spectral(es)
+!
+!             ! ubar and vbar are used here to store the mean x and y components of the vorticity
+!             ubar = svor(:, 0, 0, 1)
+!             vbar = svor(:, 0, 0, 2)
+!
+!             call diffx(es, svor(:, :, :, 1)) ! E_x
+!             call diffy(ds, cs)               ! cs = D_y
+!             !$omp parallel do private(iz)  default(shared)
+!             do iz = 0, nz
+!                svor(iz, :, :, 1) = k2l2i * (svor(iz, :, :, 1) + cs(iz, :, :))
+!             enddo
+!             !$omp end parallel do
+!
+!             call diffy(es, svor(:, :, :, 2)) ! E_y
+!             call diffx(ds, cs)               ! D_x
+!
+!             !$omp parallel do private(iz)  default(shared)
+!             do iz = 0, nz
+!                svor(iz, :, :, 2) = k2l2i * (svor(iz, :, :, 2) - cs(iz, :, :))
+!             enddo
+!             !$omp end parallel do
+!
+!             ! bring back the mean x and y components of the vorticity
+!             svor(:, 0, 0, 1) = ubar
+!             svor(:, 0, 0, 2) = vbar
 
-            call diffx(es, svor(:, :, :, 1)) ! E_x
-            call diffy(ds, cs)               ! cs = D_y
-            !$omp parallel do private(iz)  default(shared)
-            do iz = 0, nz
-               svor(iz, :, :, 1) = k2l2i * (svor(iz, :, :, 1) + cs(iz, :, :))
-            enddo
-            !$omp end parallel do
-
-            call diffy(es, svor(:, :, :, 2)) ! E_y
-            call diffx(ds, cs)               ! D_x
+            !----------------------------------------------------------
+            ! Enforce solenoidality:
+            ! we do not need to fill the internal halo grid points as
+            ! diverge calculates the x and y derivatives in spectral space where
+            ! no halo grid points are needed
 
-            !$omp parallel do private(iz)  default(shared)
-            do iz = 0, nz
-               svor(iz, :, :, 2) = k2l2i * (svor(iz, :, :, 2) - cs(iz, :, :))
-            enddo
-            !$omp end parallel do
+            ! Get source div(vortg) for solenoidal correction:
+            call divergence(vor, div)
 
-            ! bring back the mean x and y components of the vorticity
-            svor(:, 0, 0, 1) = ubar
-            svor(:, 0, 0, 2) = vbar
+            ! Compute vorticity correction grad(phi) where
```

Comment:
Should be -grad(phi).  Also, I don't see where you subtract grad(phi) from the vorticity to actually correct it.

Reply:
Sorry, I am still porting over ... this got missed.